### PR TITLE
MML-215 Datepicker

### PR DESCRIPTION
**Pull request by symphony-enrico**
See https://github.com/symphonyoss/messageml-utils/issues/207

Currently under development but most of the work is done, it can be already reviewed. Remaining:

- [x] Check max attributes length (64 and 1024)

- [x] More Junits

- [x] Test other thant simply MessageML to PresentationML

### CODE REVIEW

File: src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+  }
+
+  /**
+   * The Json for PresentationML is different from MessageML
+   * It needs to be rewritten, by adding the type attribute, based on the content
+   *
+   * @param attributeName
+   * @return
+   */
+  private String convertJsonDateToPresentationML(String attributeName) {
```
Comment (symphony-thibault) [inline src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java:190]: That's very complex :) Could you try to refactor the logic? Maybe using a Java POJO that you would then convert to JSON string

File: src/test/java/org/symphonyoss/symphony/messageml/elements/form/DatePickerTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  private String formId;
+
+  @Before
+  public void beforeEach() {
+    this.formId = "datepicker-form";
+  }
+
+  @Test
+  public void testDatePicker() throws Exception {
```
[the remark this answers is not in the document]
Comment (symphony-enrico) [reply src/test/java/org/symphonyoss/symphony/messageml/elements/form/DatePickerTest.java:25]: Sure, it is in progress

File: src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        if(jsonNode.get(FROM) != null){
+          result.append(TYPE_RANGE).append("', '")
+              .append(FROM).append("': '").append(jsonNode.get(FROM).asText()).append("', '")
+              .append(TO).append("': '").append(jsonNode.get(TO).asText()).append("'");
+        } else if(jsonNode.get(DAY) != null){
+          result.append(TYPE_DATE).append("', '")
+              .append(DAY).append("': '").append(jsonNode.get(DAY).asText()).append("'");
+        } else if(jsonNode.get(DAYS_OF_WEEK) != null){
+          result.append(TYPE_WEEKDAYS).append("', '")
+              .append(DAYS_OF_WEEK).append("': ").append(jsonNode.get(DAYS_OF_WEEK).toPrettyString());
```
[the remark this answers is not in the document]
Comment (symphony-enrico) [reply src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java:210]: asText was my first version, but it does not work:

Screen Shot 2020-11-18 at 17 57 19

Screen Shot 2020-11-18 at 17 57 06

File: src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java (hunk 1 above)
Comment (symphony-enrico) [reply src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java:190]: Yes, I thought exactly the same thing after this first implementation, it is in progress...

Comment (symphony-enrico) [reply src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java:189]: Refactored @symphony-thibault . Now the main class is much easier, even if the POJO used as a helper for the conversion is a little more than a POJO (it contains validation code too) and unfortunately there is this ugliness that I cannot fix: 

``java
result.replace("\"", "'")
``

...when converting with Jackson the POJO to a string. The problem is, single quotes are not officially supported in json format. Jackson can tolerate them for deserializing an object (if a specific mapper parameter is set!) but apparently no way to produce a json from an object with single quotes.

File: src/test/java/org/symphonyoss/symphony/messageml/elements/form/DatePickerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (symphony-enrico) [reply src/test/java/org/symphonyoss/symphony/messageml/elements/form/DatePickerTest.java]: done

**Review by symphony-thibault — DISMISSED** [review]
LGTM 👍

**Pull request merged**